torch/csrc/jit/passes/quantization.cpp

```diff
@@ -1754,22 +1754,90 @@ std::tuple<at::Tensor, at::Tensor> FoldConvBatchNorm2dHelper::
   return std::make_tuple(new_w, new_b);
 }
 
+std::vector<c10::optional<IValue>> findEPSWeightBiasConstantsOfBN(
+    const script::Module& bn) {
+  auto graph = bn.get_method("forward").graph();
+  const PatternInfo& pattern_bn = PatternInfo::parse_from_str(R"(
+      graph(%a, %weight, %bias, %running_mean, %running_var,
+          %training, %momentum, %eps, %cudnn_enabled):
+        %bn_out = aten::batch_norm(%a, %weight, %bias, %running_mean,
+            %running_var, %training, %momentum, %eps, %cudnn_enabled)
+        return (%bn_out) )");
+  const Graph& pattern_bn_graph = *pattern_bn.pattern_graph;
+  const auto& bn_vmap = pattern_bn.vmap;
+
+  const auto& matches = findPatternMatches(pattern_bn_graph, *graph);
+  if (matches.size() == 0) {
+    return std::vector<c10::optional<IValue>>(3, c10::optional<IValue>{});
+  }
+  const auto& match = matches[0];
+  std::vector<c10::optional<IValue>> eps_weight_bias_constants;
+  eps_weight_bias_constants.reserve(3);
+  eps_weight_bias_constants.emplace_back(toIValue(match.values_map.at(bn_vmap.at("eps"))));
+  eps_weight_bias_constants.emplace_back(toIValue(match.values_map.at(bn_vmap.at("weight"))));
+  eps_weight_bias_constants.emplace_back(toIValue(match.values_map.at(bn_vmap.at("bias"))));
+  return eps_weight_bias_constants;
+}
+
+bool extractOptionalBNParams(
+    const script::Module& bn,
+    ConvBNParameters& r) {
+  auto eps_weight_bias_constants = findEPSWeightBiasConstantsOfBN(bn);
+
+  if (bn.hasattr("eps")) {
+    r.bn_eps = bn.attr("eps").toDouble();
+  } else {
+    auto optional_eps = eps_weight_bias_constants[0];
+    if(!optional_eps) {
+      return false;
+    }
+    r.bn_eps = optional_eps.value().toDouble();
+  }
+  r.bn_w = at::ones_like(bn.attr("running_mean").toTensor());
+  if (bn.hasattr("weight")) {
+    if (!bn.attr("weight").isNone()) {
```

Comment:
nit: this should probably be `bn.attr("weight").isTensor()`? similar for other places

Reply:
we have a helper function: hastensor, you can use that I think

Reply:
So I did try that I remember. So I tried again to figure out what the issue was. The thing is for some traced module that when affine=False, it keep weight=bias=None attributes on BatchNorm2D, but I think the type of it somehow still remains Tensor? so isTensor returns true but it segfaults upon toTensor().
Here is an example of such a module:
```
                        module __torch__.torch.nn.modules.batchnorm.BatchNorm2d {
                          parameters {
                          }
                          attributes {
                            weight = None
                            bias = None
                            running_mean =  0.8322
                             0.9694
                             0.1845
                             0.6796
                             0.2174
                             0.3491
                             0.2117
                             0.6762
                             0.4398
                             0.2557
                             0.2554
                             0.0845
                             0.1605
                             0.1991
                             0.0240
                             0.9191
                             0.6235
                             0.7750
                             0.5440
                             0.7504
                            [ CPUDoubleType{20} ]
                            running_var =  0.6314
                             0.5476
                             0.1146
                             0.9979
                             0.5359
                             0.1544
                             0.5001
                             0.5402
                             0.7929
                             0.2334
                             0.6854
                             0.8619
                             0.9915
                             0.9608
                             0.7397
                             0.8154
                             0.8778
                             0.9027
                             0.0565
                             0.8948
                            [ CPUDoubleType{20} ]
                            num_batches_tracked = 0
                            [ CPULongType{} ]
                            training = False
                          }
                          methods {
                            method forward {
                              graph(%self : __torch__.torch.nn.modules.batchnorm.BatchNorm2d,
                                    %input.1 : Tensor):
                                %39 : Function = prim::Constant[name="batch_norm"]()
                                %38 : float = prim::Constant[value=1.0000000000000001e-05]() # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:106:40
                                %10 : bool = prim::Constant[value=0]() # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:94:11
                                %9 : bool = prim::Constant[value=1]() # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:94:29
                                %exponential_average_factor.1 : float = prim::Constant[value=0.10000000000000001]() # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:92:41
                                %16 : int = prim::Constant[value=1]() # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:97:70
                                %3 : None = prim::CallMethod[name="_check_input_dim"](%self, %input.1) # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:84:8
                                %8 : bool = prim::GetAttr[name="training"](%self)
                                %11 : bool = prim::If(%8) # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:94:11
                                  block0():
                                    -> (%9)
                                  block1():
                                    -> (%10)
                                 = prim::If(%11) # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:94:8
                                  block0():
                                    %15 : Tensor = prim::GetAttr[name="num_batches_tracked"](%self)
                                    %18 : Tensor = aten::add(%15, %16, %16) # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:97:43
                                     = prim::SetAttr[name="num_batches_tracked"](%self, %18)
                                    -> ()
                                  block1():
                                    -> ()
                                %28 : Tensor = prim::GetAttr[name="running_mean"](%self)
                                %29 : Tensor = prim::GetAttr[name="running_var"](%self)
                                %30 : None = prim::GetAttr[name="weight"](%self)
                                %31 : None = prim::GetAttr[name="bias"](%self)
                                %32 : bool = prim::GetAttr[name="training"](%self)
                                %36 : bool = prim::If(%32) # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:105:12
                                  block0():
                                    -> (%9)
                                  block1():
                                    %53 : bool = prim::Constant[value=0]()
                                    -> (%53)
                                %40 : Tensor = prim::CallFunction(%39, %input.1, %28, %29, %30, %31, %36, %exponential_average_factor.1, %38) # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:103:15
                                return (%40)

                            }
                            method _check_input_dim {
                              graph(%self : __torch__.torch.nn.modules.batchnorm.BatchNorm2d,
                                    %input.1 : Tensor):
                                %8 : None = prim::Constant() # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:247:4
                                %6 : str = prim::Constant[value="Exception"]() # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:249:12
                                %4 : int = prim::Constant[value=4]() # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:248:26
                                %3 : int = aten::dim(%input.1) # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:248:11
                                %5 : bool = aten::ne(%3, %4) # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:248:11
                                 = prim::If(%5) # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:248:8
                                  block0():
                                     = prim::RaiseException(%6) # /home/kimishpatel/local/anaconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/modules/batchnorm.py:249:12
                                    -> ()
                                  block1():
                                    -> ()
                                return (%8)

                            }
                          }
                          submodules {
                          }
                        }
```

Reply:
This looks like a bug, is this only for traced model?

Reply:
Switched to `bn.attr("weight").isTensor()`. Could not quite use `hastensor` as the structure of if conditions require splitting `hasattr` and `isTensor` check.

Reply:
I didn't see weight/bias appear in traced module actually, is this for scripted module?

Reply:
> This looks like a bug, is this only for traced model?

Found the bug :). It was me, thus removed the comment.

Reply:
> I didn't see weight/bias appear in traced module actually, is this for scripted module?

It was scripted module with BN(affine=False). Anyway the point is moot as the bug was in my code.